File: src/de/hsh/dennis/DennisGameEntryPoint.java
package de.hsh.dennis;

import common.config.WindowConfig;
import common.engine.components.game.GameEntryPoint;
import common.updates.UpdateCodes;
import common.util.Logger;
import de.hsh.dennis.controller.LevelMenu_controller;
import de.hsh.dennis.controller.MainMenu_controller;
import de.hsh.dennis.model.GameModel;
import de.hsh.dennis.model.KeyLayout;
import de.hsh.dennis.model.NpcLogic.SkinConfig;
import javafx.application.Platform;
import javafx.scene.canvas.Canvas;
import javafx.scene.input.KeyCode;

import java.util.Observable;
import java.util.Observer;


public class DennisGameEntryPoint extends GameEntryPoint {

    private DennisFxmlChanger changer;
    private GameModel gm;
    private boolean rendering = false;
    private SkinConfig.Level.Difficulty lastGameMode;


    public DennisGameEntryPoint(Observer o) {
        super(o, WindowConfig.dennis_title);
        changer = new DennisFxmlChanger(this, "view/mainMenu.fxml", new MainMenu_controller());
        gm = new GameModel();
        gm.addObserver(this);
    }

    @Override
    public void render(int fps) {
        if(gm.getFps() != fps){gm.setFps(fps);}
        if (rendering) {
            Platform.runLater( () -> {
                gm.act();
            } );
        }
    }

    @Override
    public void update(Observable o, Object arg) {

        if(arg instanceof SkinConfig.Level.Difficulty){
            lastGameMode = (SkinConfig.Level.Difficulty) arg;
            gm.setDifficulty((SkinConfig.Level.Difficulty)arg);
        }

        if (o instanceof GameModel) {
            if (arg instanceof String) {
                switch ((String)arg) {
                    case UpdateCodes.Dennis.gameLost:
                        Logger.log("!!! YOU LOSE !!!");
                        rendering = false;
                        changer.changeFxml(o, (String) arg);
                        gm.reset();
                        break;

                    case UpdateCodes.Dennis.gameWon:
                        Logger.log("!!! YOU WIN !!!");
                        rendering = false;
                        changer.changeFxml(o, (String) arg);
                        gm.reset();
                        break;
                }
            }


        } else if (arg instanceof Canvas) {
            gm.setCanvas((Canvas) arg);


        } else if (arg instanceof KeyCode) {
            if (arg == KeyLayout.Control.BREAK || arg == KeyLayout.Control.BREAK_ALT) {
                gm.triggerBreak();
                rendering = false;
                changer.changeFxml(o, arg.toString());      //no game pausing
            } else {
                gm.userInput((KeyCode) arg);
            }

        } else if (arg instanceof String) {
            if (arg.equals(UpdateCodes.Dennis.gameReady)) {
                rendering = true;
            }else if(arg.equals(UpdateCodes.Dennis.replay)){
                gm.reset();
                loadReplay();
            }else if(arg.equals(UpdateCodes.Dennis.continiue)){
                gm.unTriggerBreak();
                rendering = true;
            }else if(arg.equals("b_main_menu")){
                gm.reset();
                changer.changeFxml(o, (String) arg);
            }
            else{
                changer.changeFxml(o, (String) arg);
            }
        }
    }

    private void loadReplay() {
        switch (lastGameMode){
            case EASY:
                changer.changeFxml(new LevelMenu_controller(), "b_easy");
                break;
            case MEDIUM:
                changer.changeFxml(new LevelMenu_controller(), "b_medium");
                break;
            case HARD:
                changer.changeFxml(new LevelMenu_controller(), "b_hard");
                break;
            case NIGHTMARE:
                changer.changeFxml(new LevelMenu_controller(), "b_nightmare");
                break;
        }
    }

}

File: Games/Dennis/src/de/hsh/dennis/DennisGameEntryPoint.kt
package de.hsh.dennis

import common.config.WindowConfig
import common.engine.components.game.GameEntryPoint
import common.updates.UpdateCodes
import common.util.Logger
import de.hsh.dennis.controller.LevelMenu_controller
import de.hsh.dennis.controller.MainMenu_controller
import de.hsh.dennis.model.Difficulty
import de.hsh.dennis.model.GameModel
import de.hsh.dennis.model.KeyLayout
import de.hsh.dennis.model.Spawn.NpcHandler
import javafx.application.Platform
import javafx.scene.canvas.Canvas
import javafx.scene.input.KeyCode
import java.lang.IllegalStateException
import java.util.*

class DennisGameEntryPoint(o: Observer) : GameEntryPoint(o, WindowConfig.dennis_title) {
    private val changer: DennisFxmlChanger = DennisFxmlChanger(this, "view/mainMenu.fxml", MainMenu_controller())
    private val gm: GameModel = GameModel()
    private var rendering = false
    private var lastGameMode: Difficulty? = null

    init {
        gm.addObserver(this)
    }

    override fun render(fps: Int) {
        if (!gm.isActing) gm.printLoading()
        if (gm.fps != fps) gm.fps = fps
        if (rendering) Platform.runLater { gm.act() }
    }

    override fun update(o: Observable, arg: Any) {
        if (arg is Difficulty) {
            lastGameMode = arg
            gm.difficulty = arg
        }
        if (arg is String) {
            if (o is GameModel) when (arg) {
                UpdateCodes.Dennis.gameLost -> {
                    Logger.log("!!! YOU LOSE !!!")
                    rendering = false
                    changer.changeFxml(o, arg)
                    gm.reset()
                }
                UpdateCodes.Dennis.gameWon  -> {
                    Logger.log("!!! YOU WIN !!!")
                    rendering = false
                    changer.changeFxml(o, arg)
                    gm.reset()
                }
            }
        } else if (arg is Canvas) {
            gm.canvas = arg
            gm.npcHandler = NpcHandler(arg)
        } else if (arg is KeyCode) {
            if (arg === KeyLayout.Control.BREAK || arg === KeyLayout.Control.BREAK_ALT) {
                gm.triggerBreak()
                rendering = false
                changer.changeFxml(o, arg.toString()) //no game pausing
            } else {
                gm.userInput(arg)
            }
        } else if (arg is String) {
            when (arg) {
                UpdateCodes.Dennis.gameReady -> rendering = true
                UpdateCodes.Dennis.replay    -> {
                    gm.reset()
                    when (lastGameMode) {
                        Difficulty.EASY      -> changer.changeFxml(LevelMenu_controller(), "b_easy")
                        Difficulty.MEDIUM    -> changer.changeFxml(LevelMenu_controller(), "b_medium")
                        Difficulty.HARD      -> changer.changeFxml(LevelMenu_controller(), "b_hard")
                        Difficulty.NIGHTMARE -> changer.changeFxml(LevelMenu_controller(), "b_nightmare")
                        Difficulty.CUSTOM -> throw IllegalStateException("Custom difficulty is not implemented")
                        null                 -> throw IllegalStateException("lastGameMode is null")
                    }
                }
                UpdateCodes.Dennis.continiue -> {
                    gm.unTriggerBreak()
                    rendering = true
                }
                "b_main_menu"                -> {
                    gm.reset()
                    changer.changeFxml(o, arg)
                }
                else                         -> changer.changeFxml(o, arg)
            }
        }
    }
}
